### ISSUE TITLE
CompilerVisibleProperty semicolon escaping bug

### ISSUE THREAD

**Issue by m0sa** [dotnet/roslyn#51692]
**Version Used**: 
5.0.103

**Steps to Reproduce**:
1. In a csproj which consumes a configurable analyzer / generator, add
   ```xml
   <PropertyGroup>
     <Foo>Foo;Bar</Foo>
    <!-- I've also tried this, w/o success
      <Foo_>Foo;Bar</Foo>
      <Foo>$([MSBuild]::Escape('$(Foo_)'))</Foo>
     -->
   </PropertyGroup>
   <ItemGroup>
    <CompilerVisibleProperty Include="Foo" />
   </ItemGroup>
   ```
2. In the analyzer
    ```csharp
    AnalyzerConfigOptions.GlobalOptions.TryGetValue("build_property.Foo", out string value)
    ```

**Expected Behavior**:
`value` should contain the whole value - `"Foo;Bar"`

**Actual Behavior**:
`value` contains only `"Foo"`, `;Bar` is escaped as an [.INI comment](http://docs.editorconfig.org/en/master/editorconfig-format.html)

-----

I've dug around a bit, and the issue seems to be in how the parameters are being passed to WriteAllLines here:

https://github.com/dotnet/roslyn/blob/0aaa2ebf73a810a3e87a32dc9230af4546413976/src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets#L150

It unescapes the `_GeneratedEditorConfigFileContent` value even if it's (manually) escaped:

Binlog:
![build](https://user-images.githubusercontent.com/503702/110135968-6cae5a00-7dcf-11eb-995a-0aff750abc82.png)

Generated editorconfig:
![editorconfig content](https://user-images.githubusercontent.com/503702/110138338-2ad2e300-7dd2-11eb-8e53-e680de1f1815.png)

Analyzer execution:
![analyzer](https://user-images.githubusercontent.com/503702/110138217-06770680-7dd2-11eb-93d0-0dfa0fd12e97.png)

**Comment by sharwell** [comment]
This is effectively a duplicate of #51625, but is occurring under very different circumstances.

**Comment by m0sa** [comment]
@sharwell indeed, although I stopped looking at this point. Seems like in order for a full round-trip with special characters to work, there would need to be a well defined escaping spec (which seems to be missing now, the value is passed on as is).

There are actually 2 bugs here:

1. the msbuild escaped value is unescaped before being written to the editorconfig. In my example above, I'm manually escaping a semicolon to %3b, but the the WriteLinesToFile unescapes it. The editorconfig should contain the %3b instead of a semicolon.
2. There is no escaping happening automatically in msbuild, and there's no unescaping happening when the values are parsed here (e.g. it could get base64 encoded/decoded): https://github.com/dotnet/roslyn/blob/1b6c125cc8783de44be40dce85917edb43dc8f42/src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.cs#L219-L232

The issue I'm pointing out, though, is not a full round trip, just the fact that the msbuild-escaped value is unsecaped before being written to the file.

**Comment by jaredpar** [comment]
@mavasani, @chsienki can u all elaborate on this part? 

> Seems like in order for a full round-trip with special characters to work, there would need to be a well defined escaping spec (which seems to be missing now, the value is passed on and parsed as-is).

**Comment by m0sa** [comment]
Here's the workaround I have, so I can pass values containing arbitrary characters from csproj to analyzers:

Analyzer:

```csharp
var foo =
    context.AnalyzerConfigOptions.GlobalOptions.TryGetValue("build_property.Foo", out var foo64) 
    && !string.IsNullOrWhiteSpace(foo64)
        ? System.Text.Encoding.UTF8.GetString(System.Convert.FromBase64String(foo64))
        : null;
```


.csproj consuming (and configuring) the analyzer:

```xml
  <ItemGroup>
    <CompilerVisibleProperty Include="Foo" />
  </ItemGroup>
  <UsingTask TaskName="ToBase64" TaskFactory="RoslynCodeTaskFactory" AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.Core.dll">
    <ParameterGroup>
      <In ParameterType="System.String" Required="true" />
      <Out ParameterType="System.String" Output="true" />
    </ParameterGroup>
    <Task>
      <Code Type="Fragment" Language="C#">Out = System.Convert.ToBase64String(System.Text.Encoding.UTF8.GetBytes(In));</Code>
    </Task>
  </UsingTask>
  <Target Name="RoslynIssue51692Workaround" BeforeTargets="BeforeCompile" Condition=" '$(Foo)' != '' ">
    <ToBase64 In="$(LocalizedStringsConnectionString)">
      <Output TaskParameter="Out" PropertyName="Foo" />
    </ToBase64>
  </Target>
```

**Comment by chsienki** [comment]
@m0sa Funnily enough base64 encoding was one of the options we considered for how to route everything through the various layers correctly. 

@jaredpar The problem is that we take an msbuild value -> write that to .globalconfig -> read it into the compiler. There is an impedence mismatch between the way msbuild + .globalconfig handles escaping of values, meaning its possible to loose data. A simple example of this is multi-line values in msbuild. We'll lose everything after the initial line.

**Comment by Sebazzz** [comment]
I also found that having a newline in your compiler property causes the rest of the contents to spill over into the  `GlobalOptions` - effectively becoming visible.

**Comment by AlexeyRaga** [comment]
I am having the same (similar?) issue when using multiline text as a value.
I have a property made visible:

```xml
<CompilerVisibleProperty Include="Avro_LogicalTypes" />
```

and then use it as:

```xml
<PropertyGroup>
    <Avro_LogicalTypes>
        foo : bar,
        user-id:Contrib.Avro.CodeGen.Tests.UserId ,
        duration:System.TimeSpan
    </Avro_LogicalTypes>
</PropertyGroup>
```

When using `TryGetValue`, it ignores all the multiline text.

**Comment by jcracknell** [comment]
Can we get a note about this issue added to the appropriate sections of the source/incremental generator cookbook documents?

I've now spent ~8 hours beating my head against the wall trying to create an optional interceptor generator by checking InterceptorsNamspaces, but because I was providing `<InterceptorsNamespaces>$(InterceptorsNamespaces);MyNamespace</InterceptorsNamespaces>` I was getting no input.

**Comment by RikkiGibson** [comment]
It looks like editorconfig made a spec change which might alleviate the situation: https://spec.editorconfig.org/#no-inline-comments

> A ; or # anywhere other than at the beginning of a line does not start a comment, but is part of the text of that line. For example:
> 
> ```editorconfig
> [*.txt]
> foo = editorconfig ;)
> ```
>
> gives variable foo the value editorconfig ;) in *.txt files, not the value editorconfig.

I wonder if we should change the way we do editorconfig parsing accordingly, either unconditionally or while also providing some compat mode.
